### PR TITLE
Global Styles: Show notice in privacy settings

### PR DESCRIPTION
**Pull request by rcrdortiz**
This PR implements the UI changes necessary for displaying a notice in the privacy settings when our users are using Global Styles and are required to upgrade their plan in order to make the customization changes they've made available publicly.

I've decided not to show the notice when the site is not launched because of two main reasons. This isn't specified in the design and I think this wasn't taken into consideration, the second reason is to reduce friction when launching a site, I experimented a bit with the UI and it becomes cluttered with buttons and we aren't providing a clear path that our users should follow, which could lead to a decrease in the number of sites that get launched. 

Also, when a user launches a free site, he will already get prompted to purchase a domain and a plan.

Captura de Pantalla 2022-10-19 a las 10 12 03

#### Proposed Changes

* We now check if the user has GS in use and if we should show a notice.
* If the user qualifies, we show a notice that includes a button so that the user can upgrade his plan.

#### Testing Instructions

- [MAKE SURE TO INSTALL THE ETK OF THIS BRANCH](https://github.com/Automattic/wp-calypso/pull/69242#issuecomment-1285541319).
- MAKE SURE YOU ADD THE STICKER BLOG TO THE SITES YOU'LL WORK WITH.

#### Happy Case
* Go to the general settings screen, privacy section. http://calypso.localhost:3000/settings/general/{siteSlug}
* Make sure your site is `FREE` and that you have some GS customizations applied.
   * When the site is not launched you should not see a notice
   * When the site is launched you should see a notice.
      * Make sure the buttons on the notice work as expected.
* Remove or reset your customizations.
* You should not see the notice.

#### Mobile devices
* Repeat the happy case using a mobile viewport (just the part where you interact with the notice).

#### Test un-launched site
* Create a site and make sure to not launch it.
* Make sure it has the corresponding blog sticker.
* Go to the general settings screen, privacy section.
* You should not see the notice.

#### Test tracking event when clicking on the upgrade button.
* Go through the Happy Case
* Open tracks and search for `calypso_global_styles_gating_%`
* You should see some events named `calypso_global_styles_gating_settings_notice_upgrade_click`

#### Test network error when fetching global styles info
* Block the global styles info request in Chrome's inspector and reload.
* You should see no console errors.
* You should see no errors on the screen.
* You should see the last value cached by `useQuery`.

#### Test for regressions
* Accent color should continue to work as before. See this [PR](https://github.com/Automattic/wp-calypso/pull/68841) for testing instructions

### Test evidence
Captura de Pantalla 2022-10-20 a las 15 27 51

#### Pre-merge Checklist

- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?

Related to #

### CODE REVIEW

File: client/state/sites/hooks/use-premium-global-styles.ts [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+	id: siteId ?? 0,
+} );
+
+const getGlobalStylesInfoForSite = ( site: Site ) => {
+	if ( ! site.isCreated ) {
+		return NEW_SITE_GLOBAL_STYLES_INFO;
+	}
+
+	const globalStylesInfoEndpoint = getGSInfoEndpointForSite( site.id );
+	return wpcom.req.get( globalStylesInfoEndpoint );
```
Comment (rcrdortiz) [inline client/state/sites/hooks/use-premium-global-styles.ts]: The `useQuery` library handles success and error cases. On success, since we don't need any special mapping, we can just get away without any specific logic. If an error occurs, the `useQuery` retries the request, it does it with some sort of `exponential backoff`, but linear backoff since the timing function is f(t) = 2*t and the domain of t is [0,4]. 

So thanks to `useQuery` I can get away without needing to do basically anything.

File: client/state/sites/hooks/use-premium-global-styles.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+const mapSiteIdToSite = ( siteId: number | null ) => ( {
+	isCreated: !! siteId,
+	id: siteId ?? 0,
+} );
+
+const getGlobalStylesInfoForSite = ( site: Site ) => {
+	if ( ! site.isCreated ) {
+		return NEW_SITE_GLOBAL_STYLES_INFO;
+	}
```
Comment (rcrdortiz) [inline client/state/sites/hooks/use-premium-global-styles.ts]: The only case where we are going to need to check the feature flag is when sites aren't created. We'll need to revisit this file in the future so that we can replace the feature check with `true`, since every single site that isn't created will need to display any notices that should be displayed.

File: client/state/sites/hooks/use-premium-global-styles.ts [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+		return site?.ID ?? null;
+	} );
+
+	const site = mapSiteIdToSite( siteId );
+
+	const { data } = useQuery(
+		[ 'globalStylesInfo', site ],
+		() => getGlobalStylesInfoForSite( site ),
+		{
+			placeholderData: DEFAULT_GLOBAL_STYLES_INFO,
```
Comment (rcrdortiz) [inline client/state/sites/hooks/use-premium-global-styles.ts:68]: The `placeholderData` option returns a default value while the cache is being set, the default value isn't persisted in the actual cache. By default, we'll want to not show any notices since we can be on a paid site.

File: client/state/sites/hooks/use-premium-global-styles.ts [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+	return wpcom.req.get( globalStylesInfoEndpoint );
+};
+
+export function usePremiumGlobalStyles(): GlobalStylesInfo {
+	const params = new URLSearchParams( window.location.search );
+	const siteSlugParam = params.get( 'siteSlug' );
+	const siteIdParam = params.get( 'siteId' );
+	const selectedSiteId = useSelector( getSelectedSiteId );
+
+	// When site id is null it means that the site hasn't been created yet.
```
Comment (rcrdortiz) [inline client/state/sites/hooks/use-premium-global-styles.ts:51]: The fact that if the `siteId` is `null` means that the site isn't created isn't obvious at all, so I've added this comment. And now we have the `Site` type which should make the code more readable in regards to sites that aren't created.

File: client/my-sites/site-settings/form-general.jsx [hunk 5]
```diff
@@ -260,6 +261,12 @@ export class SiteSettingsFormGeneral extends Component {
 		} );
 	};
 
+	trackAdvancedCustomizationUpgradeClick = () => {
+		this.props.recordTracksEvent( 'calypso_global_styles_gating_settings_notice_upgrade_click', {
```
Comment (rcrdortiz) [inline client/my-sites/site-settings/form-general.jsx:265]: We'll need to add this event to Tracks.

File: apps/editing-toolkit/editing-toolkit-plugin/wpcom-global-styles/api/class-global-styles-info-rest-api.php [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+		);
+	}
+
+	/**
+	 * Checks if the user has the necessary permissions to get global styles information.
+	 *
+	 * @return bool|WP_Error
+	 */
+	public function permissions_check() {
+		if ( ! current_user_can( 'edit_theme_options' ) ) {
```
Comment (mmtr) [inline apps/editing-toolkit/editing-toolkit-plugin/wpcom-global-styles/api/class-global-styles-info-rest-api.php]: Nitpick (non-blocking): Would this cause any issue to users visiting the `/settings/general` screen without this capability? I don't know if that's possible though.

File: apps/editing-toolkit/editing-toolkit-plugin/wpcom-global-styles/api/class-global-styles-info-rest-api.php [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+class Global_Styles_Info_Rest_API extends WP_REST_Controller {
+
+	/**
+	 * Class constructor
+	 */
+	public function __construct() {
+		$this->namespace                       = 'wpcom/v2';
+		$this->rest_base                       = 'global-styles-info';
+		$this->wpcom_is_site_specific_endpoint = true;
+		$this->wpcom_is_wpcom_only_endpoint    = true;
```
Comment (mmtr) [inline apps/editing-toolkit/editing-toolkit-plugin/wpcom-global-styles/api/class-global-styles-info-rest-api.php]: Nitpick (non-blocking): This is preventing the endpoint from returning accurate data for Atomic sites, since it'll run on a `wpcom` context (not in the remote site) where the global styles CPT is not available (so `wpcom_global_styles_in_use` will always return false regardless of whether GS are being used).

This is not a blocker because we're not limiting GS on Atomic sites right now, but I think we should fix this sooner than later (not necessarily because we decide to limit GS on Atomic sites, but because this endpoint might be used for something else).

File: apps/editing-toolkit/editing-toolkit-plugin/wpcom-global-styles/api/class-global-styles-info-rest-api.php [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+		return array(
+			'globalStylesInUse'       => wpcom_global_styles_in_use(),
+			'shouldLimitGlobalStyles' => wpcom_should_limit_global_styles(),
+		);
+	}
+}
+
+if ( function_exists( 'wpcom_rest_api_v2_load_plugin' ) ) {
+	wpcom_rest_api_v2_load_plugin( 'Global_Styles_Info_Rest_API' );
+}
```
Comment (mmtr) [inline apps/editing-toolkit/editing-toolkit-plugin/wpcom-global-styles/api/class-global-styles-info-rest-api.php]: Nitpick (non-blocking): Similar as above, this only works on Simple sites. Once we change `wpcom_is_wpcom_only_endpoint` to false so it works on the Atomic context, we should update this logic as well.

File: apps/editing-toolkit/editing-toolkit-plugin/wpcom-global-styles/api/class-global-styles-info-rest-api.php [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+ * This class contains the necessary endpoints to interact with global styles outside the editor context.
+ */
+class Global_Styles_Info_Rest_API extends WP_REST_Controller {
+
+	/**
+	 * Class constructor
+	 */
+	public function __construct() {
+		$this->namespace                       = 'wpcom/v2';
+		$this->rest_base                       = 'global-styles-info';
```
Comment (mmtr) [inline apps/editing-toolkit/editing-toolkit-plugin/wpcom-global-styles/api/class-global-styles-info-rest-api.php]: Super nitpicky (non-blocking): All endpoints technically return information, so adding a `-info` suffix looks redundant. I'd probably call it just `global-styles/status` so we can use the `global-styles` path for other endpoints if needed.

File: client/my-sites/site-settings/form-general.jsx [hunk 10]
```diff
@@ -260,6 +261,12 @@ export class SiteSettingsFormGeneral extends Component {
...
+			cta_name: 'settings_site_privacy',
```
Comment (mmtr) [inline client/my-sites/site-settings/form-general.jsx:266]: Nitpick (non-blocking): Could this be renamed to `context` so it's consistent with the other events?

https://github.com/Automattic/wp-calypso/blob/c23e3520bfb9fd22aad5f8f17befd6eaccebf6ee/apps/editing-toolkit/editing-toolkit-plugin/wpcom-global-styles/notice.js#L46-L48

File: client/my-sites/site-settings/form-general.jsx [hunk 11]
```diff
@@ -698,6 +707,35 @@ export class SiteSettingsFormGeneral extends Component {
 			</div>
 		);
 	}
+
+	advancedCustomizationNotice() {
+		const { translate, siteSlug } = this.props;
+		const upgradeUrl = '/plans/' + siteSlug;
+		const siteUrl = `https://${ siteSlug }`;
```
Comment (mmtr) [inline client/my-sites/site-settings/form-general.jsx]: The site slug is not always the site URL since sites can have custom domains. See https://github.com/Automattic/wp-calypso/blob/14c651c7fea089df73d5198debfc39ad7e92ef85/client/my-sites/customer-home/main.jsx#L103

File: client/my-sites/site-settings/form-general.jsx [hunk 12]
```diff
@@ -698,6 +707,35 @@ export class SiteSettingsFormGeneral extends Component {
...
+
+		return (
+			<>
+				<div className="site-settings__advanced-customization-notice">
+					<div className="site-settings__advanced-customization-notice-cta">
+						<Gridicon icon="info-outline" />
+						<span>
+							{ translate( 'Your style changes wont be public until you upgrade your plan.' ) }
```
Comment (mmtr) [inline client/my-sites/site-settings/form-general.jsx]: Minor typo.
```suggestion
							{ translate( "Your style changes won't be public until you upgrade your plan." ) }
```

File: client/my-sites/site-settings/form-general.jsx [hunk 13]
```diff
@@ -698,6 +707,35 @@ export class SiteSettingsFormGeneral extends Component {
...
+						</span>
+					</div>
+					<div className="site-settings__advanced-customization-notice-buttons">
+						<Button href={ siteUrl }>{ translate( 'View site' ) }</Button>
```
Comment (mmtr) [inline client/my-sites/site-settings/form-general.jsx]: I think this should open a new tab since the user is leaving Calypso.
```suggestion
						<Button href={ siteUrl } target="_blank">{ translate( 'View site' ) }</Button>
```

File: client/my-sites/site-settings/style.scss [hunk 14]
```diff
@@ -625,3 +625,38 @@ button.site-settings__general-settings-set-guessed-timezone.button.is-borderless
 	margin-top: -16px;
 	margin-bottom: 16px;
 }
+
+.site-settings__advanced-customization-notice {
+	display: flex;
+	padding: 20px;
+	background-color: #f4f4f4;
```
Comment (mmtr) [inline client/my-sites/site-settings/style.scss]: I think this should use a variable from the color scheme: https://github.com/Automattic/wp-calypso/blob/trunk/packages/calypso-color-schemes/src/shared/color-schemes/_default.scss

Maybe `--color-neutral-0`?

File: client/my-sites/site-settings/style.scss [hunk 15]
```diff
@@ -625,3 +625,38 @@ button.site-settings__general-settings-set-guessed-timezone.button.is-borderless
...
+	align-items: center;
+	justify-content: space-between;
+
+	.site-settings__advanced-customization-notice-cta {
+		display: flex;
+		align-items: center;
+	}
+
+	svg.gridicon.gridicons-info-outline {
+		color: #d38d16;
```
Comment (mmtr) [inline client/my-sites/site-settings/style.scss]: I think this should use a variable from the color scheme: https://github.com/Automattic/wp-calypso/blob/trunk/packages/calypso-color-schemes/src/shared/color-schemes/_default.scss

Maybe `--color-warning-30`?

File: client/my-sites/site-settings/style.scss [hunk 16]
```diff
@@ -625,3 +625,38 @@ button.site-settings__general-settings-set-guessed-timezone.button.is-borderless
...
+	.site-settings__advanced-customization-notice-buttons {
+		display: flex;
+
+		.button.is-primary {
+			margin-left: 10px;
+
+		}
+	}
+
+	@media (max-width: 1100px) {
```
Comment (mmtr) [inline client/my-sites/site-settings/style.scss]: Any reason why this is not using any of the existing breakpoints? See https://wpcalypso.wordpress.com/devdocs/docs/coding-guidelines/css.md#media-queries or https://github.com/Automattic/wp-calypso/blob/trunk/client/assets/stylesheets/shared/mixins/_breakpoints.scss

File: client/state/sites/hooks/use-premium-global-styles.ts (hunk 1 above)
Comment (mmtr) [inline client/state/sites/hooks/use-premium-global-styles.ts]: Nitpick (non-blocking): I personally find this very hard to parse mentally. I prefer something like the below so we can effectively remove some functions like `mapSiteIdToSite` or `getGSInfoEndpointForSite` since they are not needed anywhere else.

```suggestion
const getGlobalStylesInfoForSite = ( siteId: number | null ) => {
	if ( ! siteId ) {
		return {
			shouldLimitGlobalStyles: isEnabled( 'limit-global-styles' ),
			globalStylesInUse: false,
		}
	}

	return wpcom.req.get( {
		path: `sites/${ siteId }/global-styles-info`,
		apiNamespace: 'wpcom/v2/',
	} );
```

**Review by mmtr — COMMENTED** [review]
Nice work! Overall this tests very well, but I think we should address a few things before shipping.

- You need to consider that the copy of the notice might be longer in other languages and wrapped in two lines. The layout is a bit off in such cases right now:
Screen Shot 2022-10-21 at 13 22 33

- I think we should add the endpoint in a separate PR so we can deploy a new ETK version before deploying the Calypso changes that start fetching data from it

- Looks like we make a recurrent request to the endpoint. That's not usually our approach when fetching data from an endpoint due to performance reasons (the only exception is when the fetched data is very critical and needs to be reflected in real time, but this not the case). The data should be loaded only once when the component first renders. I realize this can cause issues like the notice not showing up if you have the settings page open and are editing the Global Styles in a separate tab, but that's edge-case scenario and it's fine to don't cover it.

Screen Shot 2022-10-21 at 13 21 34

File: client/state/sites/hooks/use-premium-global-styles.ts (hunk 1 above)
Comment (rcrdortiz) [reply client/state/sites/hooks/use-premium-global-styles.ts]: For me, it's the other way around XD. I personally find your suggestion much harder to follow since I need to parse everything that the code is doing instead of actually reading the code.

File: apps/editing-toolkit/editing-toolkit-plugin/wpcom-global-styles/api/class-global-styles-info-rest-api.php (hunk 7 above)
Comment (rcrdortiz) [reply apps/editing-toolkit/editing-toolkit-plugin/wpcom-global-styles/api/class-global-styles-info-rest-api.php]: We can tackle this once this endpoint is used form something else then.

File: apps/editing-toolkit/editing-toolkit-plugin/wpcom-global-styles/api/class-global-styles-info-rest-api.php (hunk 6 above)
Comment (rcrdortiz) [reply apps/editing-toolkit/editing-toolkit-plugin/wpcom-global-styles/api/class-global-styles-info-rest-api.php]: They would not see the notice if they don't have the capability.

File: apps/editing-toolkit/editing-toolkit-plugin/wpcom-global-styles/api/class-global-styles-info-rest-api.php (hunk 8 above)
Comment (rcrdortiz) [reply apps/editing-toolkit/editing-toolkit-plugin/wpcom-global-styles/api/class-global-styles-info-rest-api.php]: Same as the other comment, if this is ever needed we can iterate.

File: client/my-sites/site-settings/form-general.jsx (hunk 10 above)
Comment (rcrdortiz) [reply client/my-sites/site-settings/form-general.jsx:266]: As it is right now it's consistent with other `recordTracksEvent` found in `form-general.jsx`.

**Comment by rcrdortiz** [comment]
> I think we should add the endpoint in a separate PR so we can deploy a new ETK version before deploying the Calypso changes that start fetching data from it

No need to do this. `useQuery` will use the default values and not show the notice until the ETK endpoint is live, which shouldn't be a few minutes after at most.

**Comment by rcrdortiz** [comment]
> Looks like we make a recurrent request to the endpoint. That's not usually our approach when fetching data from an endpoint due to performance reasons (the only exception is when the fetched data is very critical and needs to be reflected in real time, but this not the case). The data should be loaded only once when the component first renders. I realize this can cause issues like the notice not showing up if you have the settings page open and are editing the Global Styles in a separate tab, but that's edge-case scenario and it's fine to don't cover it.

Exactly how are you making it make more than one request? I'm unable to reproduce it. Maybe what you're experiencing is related to how `useQuery` handles state internally 🤔, with normal usage of the settings screen, I can't seem to reproduce it. 

https://user-images.githubusercontent.com/1989914/197204758-e0ccfa3c-a4e8-4108-977b-69b4901e7868.mp4

**Comment by mmtr** [comment]
> No need to do this. `useQuery` will use the default values and not show the notice until the ETK endpoint is live.

I got a WSOD if the endpoint is not live:

Screen Shot 2022-10-21 at 15 17 35

> Exactly how are you making it make more than one request? I'm unable to reproduce it.

Oh sorry, I thought it was a recurring request, but apparently there is a new request every time I come back to the tab after switching to a different tab:

https://user-images.githubusercontent.com/1233880/197206782-4440a3cb-986f-4d97-bfe0-78588dc0b469.mov

It's not a big deal, but I don't think that's the normal behavior we have in other components.

**Comment by rcrdortiz** [comment]
> It's not a big deal, but I don't think that's the normal behavior we have in other components.

Do you happen to know how to avoid that? It has to be related to `useQuery` but I can't seem to find anything about it.

File: client/state/sites/hooks/use-premium-global-styles.ts (hunk 1 above)
Comment (mmtr) [reply client/state/sites/hooks/use-premium-global-styles.ts]: It's a matter of getting used to it :D I found it confusing at the beginning too, but ended up paying off. I recommend you to take a look at this post: p7H4VZ-2av-p2

> Don’t wrap functions with new functions that opaquely pass along its parameters underneath. Be cautious of places where we’re adding one or two additional values onto the underlying function calls. These add indirection, meaning we have to jump through more code in order to understand them than is necessary.

> Don’t wrap APIs with more-constrained wrappers around those APIs. These tend to become quickly insufficient as people often want access to more of the underlying API than we initially assume they might

> Don’t split linear procedures into arbitrary linear sub-procedures to separate side-effects.

**Comment by rcrdortiz** [comment]
> I got a WSOD if the endpoint is not live:

Got it now! I had to wait for the `useQuery` cache to expire. It should be working now.

**Comment by mmtr** [comment]
> Do you happen to know how to avoid that? It has to be related to `useQuery` but I can't seem to find anything about it.

Looks like we have a `refetchOnWindowFocus` option to configure it: https://react-query-v2.tanstack.com/reference/useQuery

**Comment by rcrdortiz** [comment]
> > Do you happen to know how to avoid that? It has to be related to `useQuery` but I can't seem to find anything about it.
> 
> Looks like we have a `refetchOnWindowFocus` option to configure it: https://react-query-v2.tanstack.com/reference/useQuery

Oh, nice find! I've added it to our `useQuery` options.

File: apps/editing-toolkit/editing-toolkit-plugin/wpcom-global-styles/api/class-global-styles-info-rest-api.php (hunk 9 above)
Comment (rcrdortiz) [reply apps/editing-toolkit/editing-toolkit-plugin/wpcom-global-styles/api/class-global-styles-info-rest-api.php]: I've changed it to `global-styles/status` since it does make more sense from a semantic point of view. The only problem I see is that we aren't being consistent on path naming since we have some endpoints that don't follow this pattern within the global-styles domain.

File: client/my-sites/site-settings/style.scss (hunk 14 above)
Comment (rcrdortiz) [reply client/my-sites/site-settings/style.scss]: Changed, thanks for pointing to the file!

File: client/my-sites/site-settings/style.scss (hunk 15 above)
Comment (rcrdortiz) [reply client/my-sites/site-settings/style.scss]: Changed, thanks for pointing to the file!

File: client/my-sites/site-settings/style.scss (hunk 16 above)
Comment (rcrdortiz) [reply client/my-sites/site-settings/style.scss]: Changed, thanks for pointing to the file!

**Comment by rcrdortiz** [comment]
> You need to consider that the copy of the notice might be longer in other languages and wrapped in two lines. The layout is a bit off in such cases right now

This should be working now for longer translated texts.

File: client/my-sites/site-settings/form-general.jsx (hunk 11 above)
Comment (rcrdortiz) [reply client/my-sites/site-settings/form-general.jsx]: Changed to use the `URL` attribute of the selected site.

**Review by mmtr — COMMENTED** [review]
Thanks for addressing all the feedback!

Looks like there is a regression in the notices showed during onboarding.

For example, `/setup/designSetup?siteSlug=<SITE_SLUG>` used to display a premium badge on non-default variations but it's missing:
 
Screen Shot 2022-10-21 at 17 41 15

Also, when going to `/setup/newsletterSetup?flow=newsletter` there is a request to the API endpoint despite the site not being created. Where is that siteId coming from?

Screen Shot 2022-10-21 at 17 43 54

**Comment by rcrdortiz** [comment]
> Thanks for addressing all the feedback!
> 
> Looks like there is a regression in the notices showed during onboarding.
> 
> For example, `/setup/designSetup?siteSlug=<SITE_SLUG>` used to display a premium badge on non-default variations but it's missing:
> 
> Screen Shot 2022-10-21 at 17 41 15
> 
> Also, when going to `/setup/newsletterSetup?flow=newsletter` there is a request to the API endpoint despite the site not being created. Where is that siteId coming from?
> 
> Screen Shot 2022-10-21 at 17 43 54

I can't reproduce the regression you mentioned. Could you share what steps you're following?

https://user-images.githubusercontent.com/1989914/197278039-d2eb4a3e-1958-48b7-9644-373dd6643171.mp4


For the second thing you mention, there seems to be a bug with `useSelector( getSelectedSiteId );` or with the actual newsletterSetup flow, the selected site isn't getting cleaned up from the store when navigating to `/setup/newsletterSetup?flow=newsletter` so the logic we have for determining the siteId is returning the last siteId (since that is what we have in the store). Is it possible to navigate to the newsletter flow from within Calypso? Or is this something that won't happen in real-life scenarios?

**Comment by mmtr** [comment]
> I can't reproduce the regression you mentioned. Could you share what steps you're following?

I just created a new site, added the blog sticker and started the design step during the site setup process. However, I cannot reproduce it now, so maybe I was half-sandboxed, or maybe it was a caching issue. I guess it's fine to ignore for now.

> Is it possible to navigate to the newsletter flow from within Calypso? Or is this something that won't happen in real-life scenarios?

No idea how the newsletter flow starts, but I think it's something that can happen in a real-time scenario since we cannot guarantee that the user didn't visit Calypso previously (even if the flows starts outside of Calypso).

I think the problem is that all these onboarding flows use a different way of storing site data and completely ignores what's in the traditional state. I wonder if there is any selector that tells us whether the user is onboarding so we can discard whatever `getSelectedSiteId` returns.

**Comment by mmtr** [comment]
> I think the problem is that all these onboarding flows use a different way of storing site data and completely ignores what's in the traditional state. I wonder if there is any selector that tells us whether the user is onboarding so we can discard whatever `getSelectedSiteId` returns.

Maybe `select( ONBOARD_STORE ).getState()`?

**Comment by rcrdortiz** [comment]
> > I think the problem is that all these onboarding flows use a different way of storing site data and completely ignores what's in the traditional state. I wonder if there is any selector that tells us whether the user is onboarding so we can discard whatever `getSelectedSiteId` returns.
> 
> Maybe `select( ONBOARD_STORE ).getState()`?

Thanks for the insight! It kinda worked, we also need to check for some other conditions like not having site query args. It should be working now.

File: client/state/sites/hooks/use-premium-global-styles.ts [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+	isCreated: boolean;
+	id: number;
+};
+
+type Onboard = {
+	stepProgress?: {
+		count: number;
+		progress: number;
+	};
+};
```
Comment (mmtr) [inline client/state/sites/hooks/use-premium-global-styles.ts]: I think it's better if you reuse the State type exported by the package:

https://github.com/Automattic/wp-calypso/blob/fb76957541c12593dbb56e14e9fbcbe4a45a9ca1/packages/data-stores/src/onboard/index.ts#L10

That way we don't need to make changes here if the type changes in the future

File: client/state/sites/hooks/use-premium-global-styles.ts [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+const getGlobalStylesInfoForSite = ( site: Site ): GlobalStylesStatus => {
+	if ( ! site.isCreated ) {
+		return NEW_SITE_GLOBAL_STYLES_INFO;
+	}
+
+	const globalStylesInfoEndpoint = getGSInfoEndpointForSite( site.id );
+	return wpcom.req
+		.get( globalStylesInfoEndpoint )
+		.then( mapResponseToGlobalStylesStatus )
+		.catch( () => DEFAULT_GLOBAL_STYLES_INFO );
```
Comment (mmtr) [inline client/state/sites/hooks/use-premium-global-styles.ts]: Feel free to ignore this.

Just a reminder that this code is adding indirection since I need to remember 3 functions now (they were 2 in a previous commit) or go back to the lines above in order to understand the behavior.

This level of indirection is ok when you want to hide the implementation details of a very complex or long function, but in this case the functions are not complex or long at all, so you're forcing developers to make jumps when reading code.

Anyway, this is only an observation to be mindful about these abstractions, it's not a blocker.

File: client/state/sites/hooks/use-premium-global-styles.ts [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+	const selectedSiteId = useSelector( getSelectedSiteId );
+	const onboard = select( ONBOARD_STORE ).getState();
+
+	// When site id is null it means that the site hasn't been created yet.
+	const siteId = useSelector( ( state ) => {
+		if ( isOnboarding( onboard ) && ! siteIdParam && ! siteSlugParam ) {
+			return null;
+		}
+
+		const siteIdOrSlug = selectedSiteId ?? siteIdParam ?? siteSlugParam;
```
Comment (mmtr) [inline client/state/sites/hooks/use-premium-global-styles.ts]: I think it'd be safer to also ignore the `selectedSiteId` if `siteIdParam` or `siteSlugParam` are present when the user is onboarding. Otherwise the hook might use an incorrect site ID (a _cached_ `selectedSiteId`) during an onboarding process that explicitly defines a different site ID (via  `siteIdParam`).

```suggestion
		const siteIdOrSlug = isOnboarding( onboard ) ? siteIdParam ?? siteSlugParam : selectedSiteId;
```

**Review by mmtr — COMMENTED** [review]
Pretty close!!

File: client/state/sites/hooks/use-premium-global-styles.ts (hunk 18 above)
Comment (rcrdortiz) [reply client/state/sites/hooks/use-premium-global-styles.ts]: The benefit of having semantics instead of actual code is that you don't need to read how something is implemented to know what it's doing. In this specific case, you can see that we are mapping the API response to GlobalStylesStatus without reading how it's implemented (you can choose to read the implementation but aren't forced to). 

With good implementations of indirection, you don't need to remember three functions, you also don't need to navigate to understand the behavior, you just need to read 3 lines of code instead of 60. The words used should be enough to explain what is going on. Indirection could be a problem when it adds no semantic value to the code.

As an example look at the way we're determining the site's information, you need to read 15 lines of code, and understand them before you can establish that we're determining if there is a siteId, and if not, the site hasn't been created. If we have just used a function here we would just need to read one line of code, and the responsibilities of the main hook would also be reduced in scope.

I doubt we will ever reach an agreement since this is completely subjective, that's why I usually don't comment on things like this in your code reviews, I know I'm not going to be able to show you/convince you of the value I see in it, the same way you won't be able to convince me the value of de-structured code.

Comment (mmtr) [reply client/state/sites/hooks/use-premium-global-styles.ts]: > I know I'm not going to be able to show you/convince you of the value I see in it, the same way you won't be able to convince me the value of de-structured code.

I was convinced at some point! At the beginning I preferred to abstract/hide the implementation details, but then someone gave me feedback about not hiding them and I re-considered my guidelines. Which is why I gave this feedback to you now. But it's ok if you think you'll never be convinced. I'll stop providing this kind of feedback from now on.

File: client/state/sites/hooks/use-premium-global-styles.ts (hunk 19 above)
Comment (rcrdortiz) [reply client/state/sites/hooks/use-premium-global-styles.ts]: Done

File: client/state/sites/hooks/use-premium-global-styles.ts (hunk 18 above)
Comment (rcrdortiz) [reply client/state/sites/hooks/use-premium-global-styles.ts]: I see, even though I'll never be convinced about that particular way of coding being correct or superior in any way (promotes unmaintainable, non-reusable, untestable code, and increases the cognitive load for understanding it), I can adapt to it.

I've changed the code.

File: client/state/sites/hooks/use-premium-global-styles.ts (hunk 17 above)
Comment (rcrdortiz) [reply client/state/sites/hooks/use-premium-global-styles.ts]: No need now with the latest refactor.

File: client/my-sites/site-settings/form-general.jsx [hunk 20]
```diff
@@ -698,6 +707,36 @@ export class SiteSettingsFormGeneral extends Component {
...
+						<span>
+							{ translate( "Your style changes won't be public until you upgrade your plan." ) }
+						</span>
+					</div>
+					<div className="site-settings__advanced-customization-notice-buttons">
+						<Button href={ selectedSite.URL } target="_blank">
+							{ translate( 'View site' ) }
+						</Button>
+						<Button
+							className="is-primary"
```
Comment (mmtr) [inline client/my-sites/site-settings/form-general.jsx:729]: Nitpick (non-blocking): The same can be accomplished by setting a `primary` prop.
```suggestion
							primary
```

**Review by mmtr — APPROVED** [review]
Works great!

File: client/state/sites/hooks/use-premium-global-styles.ts (hunk 18 above)
Comment (mmtr) [reply client/state/sites/hooks/use-premium-global-styles.ts]: Already discussed privately on Slack, but also wanted to leave a comment here to close the loop. It was not my intention to impose a code style, but to bring up a different perspective for you to consider. If you don't see value in that perspective, it's fine to keep the implementation with the previous approach. This is something subjective as you pointed out, so you're not expected to disagree and commit.

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7594428

Thank you @rcrdortiz for including a screenshot in the description! This is really helpful for our translators.

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7594428).
